File: app/src/main/java/com/compuspulse/viewmodel/CategoryViewModel.java
package com.compuspulse.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.compuspulse.data.remote.model.Article;
import com.compuspulse.data.remote.model.ProjectCategory;
import com.compuspulse.data.repository.CategoryRepository;
import com.compuspulse.data.repository.RepositoryCallback;

import java.util.ArrayList;
import java.util.List;

public class CategoryViewModel extends ViewModel {

    private final CategoryRepository repository = new CategoryRepository();
    private final MutableLiveData<List<ProjectCategory>> categoryList = new MutableLiveData<>(new ArrayList<>());
    private final MutableLiveData<List<Article>> articleList = new MutableLiveData<>(new ArrayList<>());
    private final MutableLiveData<String> errorMessage = new MutableLiveData<>();
    private final MutableLiveData<Boolean> loading = new MutableLiveData<>(false);

    public LiveData<List<ProjectCategory>> getCategoryList() {
        return categoryList;
    }

    public LiveData<List<Article>> getArticleList() {
        return articleList;
    }

    public LiveData<String> getErrorMessage() {
        return errorMessage;
    }

    public LiveData<Boolean> getLoading() {
        return loading;
    }

    public void loadCategories() {
        loading.setValue(true);
        repository.fetchProjectCategories(new RepositoryCallback<List<ProjectCategory>>() {
            @Override
            public void onSuccess(List<ProjectCategory> data) {
                categoryList.setValue(data);
                loading.setValue(false);
                if (data != null && !data.isEmpty()) {
                    loadByCategory(data.get(0).getId());
                }
            }

            @Override
            public void onError(String message) {
                errorMessage.setValue(message);
                loading.setValue(false);
            }
        });
    }

    public void loadByCategory(int categoryId) {
        loading.setValue(true);
        repository.fetchArticlesByCategory(categoryId, new RepositoryCallback<List<Article>>() {
            @Override
            public void onSuccess(List<Article> data) {
                articleList.setValue(data == null ? new ArrayList<>() : data);
                loading.setValue(false);
            }

            @Override
            public void onError(String message) {
                articleList.setValue(new ArrayList<>());
                errorMessage.setValue(message);
                loading.setValue(false);
            }
        });
    }
}

File: app/src/main/java/com/compuspulse/viewmodel/CategoryViewModel.kt
package com.compuspulse.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.compuspulse.data.remote.model.Article
import com.compuspulse.data.remote.model.ProjectCategory
import com.compuspulse.data.repository.CategoryRepository
import kotlinx.coroutines.launch

class CategoryViewModel : ViewModel() {

    private val repository = CategoryRepository()
    private val categoryList = MutableLiveData<List<ProjectCategory>>(ArrayList())
    private val articleList = MutableLiveData<List<Article>>(ArrayList())
    private val errorMessage = MutableLiveData<String>()
    private val loading = MutableLiveData(false)

    fun getCategoryList(): LiveData<List<ProjectCategory>> = categoryList

    fun getArticleList(): LiveData<List<Article>> = articleList

    fun getErrorMessage(): LiveData<String> = errorMessage

    fun getLoading(): LiveData<Boolean> = loading

    fun loadCategories() {
        loading.value = true
        viewModelScope.launch {
            repository.fetchProjectCategories()
                .onSuccess { data ->
                    categoryList.value = data
                    loading.value = false
                    if (data.isNotEmpty()) {
                        loadByCategory(data[0].getId())
                    }
                }
                .onFailure { throwable ->
                    errorMessage.value = throwable.message ?: "分类加载失败"
                    loading.value = false
                }
        }
    }

    fun loadByCategory(categoryId: Int) {
        loading.value = true
        viewModelScope.launch {
            repository.fetchArticlesByCategory(categoryId)
                .onSuccess { data ->
                    articleList.value = data
                    loading.value = false
                }
                .onFailure { throwable ->
                    articleList.value = ArrayList()
                    errorMessage.value = throwable.message ?: "分类文章加载失败"
                    loading.value = false
                }
        }
    }
}
